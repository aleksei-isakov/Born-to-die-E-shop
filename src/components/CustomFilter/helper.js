export const ICON_WIDTH = '50%';

export const SELECTED_OPTIONS_KEYS = {
  NEW_TO_OLD: 'NEW_TO_OLD',
  OLD_TO_NEW: 'OLD_TO_NEW',
  CHEAP_TO_EXPENSIVE: 'CHEAP_TO_EXPENSIVE',
  EXPENSIVE_TO_CHEAP: 'EXPENSIVE_TO_CHEAP'
};
