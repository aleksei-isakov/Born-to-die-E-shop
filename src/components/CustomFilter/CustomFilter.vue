<template>
  <div class="filter-wrapper">
    <div class="filter-block">
      <div
        v-for="(iconPath, i) in iconsPath"
        :key="i"
        class="filter-block__item"
        :class="{ active: selectedIconPath === iconPath }"
        @click="onClickSelectIconPath(iconPath)"
      >
        <BaseCustomIcon :icon="iconPath" :width="iconWidth" />
      </div>
    </div>

    <div class="custom-select-block">
      <p class="custom-select-block__title">Filter</p>
      <div class="custom-select">
        <div
          class="custom-select__selected"
          @click.stop="onClickChangeOptionsVisibility"
        >
          {{ activeOption }}
          <p>
            <i
              class="arrow"
              :class="{
                arrow__up: isOptionsVisible,
                arrow__down: !isOptionsVisible
              }"
            ></i>
          </p>
        </div>

        <div v-if="isOptionsVisible" class="custom-select__options">
          <p
            v-for="(option, i) in options"
            :key="i"
            class="custom-select__option"
            @click="onClickSelectOption(option.value)"
          >
            {{ option.name }}
          </p>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import { BaseCustomIcon } from '@/base_components';
import { ICON_WIDTH, SELECTED_OPTIONS_KEYS } from './helper';

export default {
  name: 'CustomFilter',

  components: {
    BaseCustomIcon
  },

  props: {
    selectedIconPath: {
      type: String,
      default: ''
    }
  },

  data() {
    return {
      isOptionsVisible: false,
      iconsPath: ['menu_filter_column', 'menu_filter_grid'],
      iconWidth: ICON_WIDTH,
      options: [
        {
          value: SELECTED_OPTIONS_KEYS.NEW_TO_OLD,
          name: 'From new to old'
        },
        {
          value: SELECTED_OPTIONS_KEYS.OLD_TO_NEW,
          name: 'From old to new'
        },
        {
          value: SELECTED_OPTIONS_KEYS.CHEAP_TO_EXPENSIVE,
          name: 'From cheap to expensive'
        },
        {
          value: SELECTED_OPTIONS_KEYS.EXPENSIVE_TO_CHEAP,
          name: 'From expensive to cheap'
        }
      ],
      selectedOption: SELECTED_OPTIONS_KEYS.NEW_TO_OLD
    };
  },

  computed: {
    activeOption() {
      return this.options.find(({ value }) => value === this.selectedOption)
        ?.name;
    }
  },

  mounted() {
    document.addEventListener('click', this.onHideSelect);
  },

  beforeDestroy() {
    document.removeEventListener('click', this.onHideSelect);
  },

  methods: {
    onClickSelectIconPath(iconPath) {
      this.$emit('click', iconPath);
    },

    onClickSelectOption(value) {
      this.$emit('onClickSelectOption', value);
      this.selectedOption = value;
      this.isOptionsVisible = false;
    },

    onClickChangeOptionsVisibility() {
      this.isOptionsVisible = !this.isOptionsVisible;
    },

    onHideSelect() {
      this.isOptionsVisible = false;
    }
  }
};
</script>

<style lang="scss" scoped>
@import '@/scss/CustomVariables.scss';

* {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
}

.filter-wrapper {
  display: flex;
  justify-content: flex-end;
  align-items: flex-end;
  height: 50px;
  color: $font-color-title;
  padding: 0 15vw;

  @media screen and (max-width: $tablet-size) {
    justify-content: center;
  }
}

.filter-wrapper p {
  margin-bottom: 0;
}

.filter-block {
  display: flex;
  border: 2px solid $light-border-color;
  border-radius: 10px;
  margin-right: 30px;
  @media screen and (max-width: $tablet-size) {
    display: none;
  }
  &__item {
    display: flex;
    justify-content: center;
    align-items: center;
    width: 40px;
    height: 40px;
    border-radius: 5px;
    cursor: pointer;
  }
  &__icon {
    width: 50%;
    height: auto;
  }
}

.active {
  background-color: $light-border-color;
}

.custom-select-block__title {
  text-align: left;
  color: $font-color-text;
}

.custom-select {
  display: block;
  position: relative;
  width: 280px;
  border: 1px solid $light-border-color;
  border-radius: 5px;
  text-align: left;
  cursor: pointer;
  &__selected {
    position: relative;
    display: flex;
    justify-content: flex-start;
    align-items: center;
    padding: 0px 15px;
    width: 100%;
    height: 30px;
    & p {
      position: absolute;
      right: 10px;
    }
  }
  &__options {
    position: absolute;
    margin-top: 4px;
    padding: 0px 15px;
    width: 100%;
    height: 117px;
    border: 1px solid $light-border-color;
    border-radius: 5px;
    background-color: $white;
  }
  &__option {
    line-height: 27px;
  }
  &__option:hover {
    background-color: $light-border-color;
    transition: $transition;
  }
  .arrow {
    border: solid black;
    border-width: 0 1.5px 1.5px 0;
    display: inline-block;
    padding: 3px;
    &__up {
      transform: rotate(-135deg);
      -webkit-transform: rotate(-135deg);
    }
    &__down {
      margin-bottom: 3px;
      transform: rotate(45deg);
      -webkit-transform: rotate(45deg);
    }
  }
}
</style>
