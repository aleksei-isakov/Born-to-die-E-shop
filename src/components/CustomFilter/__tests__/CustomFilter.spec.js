import { shallowMount } from '@vue/test-utils';
import CustomFilter from '../CustomFilter.vue';

let wrapper;
const testSelectedIconPath = 'menu_filter_column';

beforeEach(() => {
  wrapper = shallowMount(CustomFilter, {
    propsData: {
      selectedIconPath: testSelectedIconPath
    }
  });
});

afterEach(() => {
  wrapper.destroy();
});

describe('FilterComponent', () => {
  test('is a Vue instance', () => {
    expect(wrapper.isVueInstance).toBeTruthy();
  });

  test('should return correct props', () => {
    expect(wrapper.props().selectedIconPath).toBe(testSelectedIconPath);
  });

  test('should open/close custom select and show/hide the options', async () => {
    const selectField = wrapper.find('.custom-select__selected');

    expect(wrapper.contains('.custom-select__selected')).toBe(true);

    await selectField.trigger('click');
    expect(wrapper.contains('.custom-select__options')).toBe(true);

    await selectField.trigger('click');
    expect(wrapper.contains('.custom-select__options')).toBe(false);
  });

  test('filter items should be clickable ', async () => {
    const filterItem = wrapper.findAll('.filter-block__item');

    await filterItem.trigger('click');
    expect(wrapper.findAll('.filter-block__item').length).toBe(2);
  });

  test('first filter element should have active class but second shoud not have', () => {
    const filterItem = wrapper.findAll('.filter-block__item');

    expect(filterItem.at(0).findAll('.active').length).toBe(1);
    expect(filterItem.at(1).findAll('.active').length).toBe(0);
  });

  test('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
