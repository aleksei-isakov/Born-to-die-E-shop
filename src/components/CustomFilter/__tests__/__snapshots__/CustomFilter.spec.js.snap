// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterComponent should match snapshot 1`] = `
<div class="filter-wrapper">
  <div class="filter-block">
    <div class="filter-block__item active">
      <customicon-stub icon="menu_filter_column" width="50%"></customicon-stub>
    </div>
    <div class="filter-block__item">
      <customicon-stub icon="menu_filter_grid" width="50%"></customicon-stub>
    </div>
  </div>
  <div>
    <p class="title">Filter</p>
    <div class="custom-select">
      <div class="custom-select__selected">
        From new to old
        <p><i class="arrow arrow__down"></i></p>
      </div>
      <!---->
    </div>
  </div>
</div>
`;
